jina/types/arrays/memmap.py

```diff
@@ -150,39 +164,121 @@ def append(self, doc: 'Document', flush: bool = True) -> None:
                 ],
             ).tobytes()
         )
-        self._header_map[doc.id] = (len(self._header_map), p, r, r + l)
+        if idx is None:
+            self._header_map[doc.id] = (self.last_header_entry, p, r, r + l)
+            self.last_header_entry = self.last_header_entry + 1
+        else:
+            self._header_map[doc.id] = (idx, p, r, r + l)
+            self._header.seek(0, 2)
         self._start = p + r + l
         self._body.write(value)
         if flush:
             self._header.flush()
             self._body.flush()
+        if update_buffer:
+            result = self.buffer_pool.add_or_update(doc.id, doc)
+            if result:
+                _key, _doc = result
+                self.update(_doc, self._str2int_id(_key), update_buffer=False)
+
+    def append(
+        self, doc: 'Document', flush: bool = True, update_buffer: bool = True
+    ) -> None:
+        """
+        Append :param:`doc` in :class:`DocumentArrayMemmap`.
 
-    def __getitem__(self, key: Union[int, str]) -> 'Document':
-        if isinstance(key, str):
-            pos_info = self._header_map[key]
-            _, p, r, l = pos_info
-            with mmap.mmap(self._body_fileno, offset=p, length=l) as m:
-                return Document(m[r:])
-        elif isinstance(key, int):
-            return self[self._int2str_id(key)]
-        else:
-            raise TypeError(f'`key` must be int or str, but receiving {key!r}')
+        :param doc: The doc needs to be appended.
+        :param update_buffer: If set, update the buffer.
+        :param flush: If set, then flush to disk on done.
+        """
+        self._update_or_append(doc, flush=flush, update_buffer=update_buffer)
+
+    def update(
```

Comment:
Are we also sure we want this to be part of the interface?

Reply:
made it protected